### ISSUE TITLE
`FailureApp`s `script_name: nil` breaks route generation within mounted engines

### ISSUE THREAD

**Issue by jvdp** [plataformatec/devise#3705]
Since upgrading to Devise 3.5.2, the `Devise::FailureApp` from within a Rails engine no longer generates routes with the path it is mounted under.

The breaking change was made in https://github.com/plataformatec/devise/commit/aa675f7f6611de5d40114aa98153f3f0c0b3e8d5, which always passes a `script_name` (possibly nil) to the url generator. This is problematic because the mount point path is only computed when `script_name` is not passed to `options`: https://github.com/rails/rails/blob/v4.2.3/actionpack/lib/action_dispatch/routing/mapper.rb#L631

I'm not sure how to fix this (this is all crazy complicated code), but perhaps there's a better fix for #3643 than forcing `path_name` to be passed to `url_for`.

**Comment by josevalim** [comment]
Which Rails version are you using?

**Comment by jvdp** [comment]
Rails 4.2.3.

**Comment by josevalim** [comment]
@jvdp so apparently if we don't set it, there is a infinite loop. If we set it, we break engines. :(

**Comment by jvdp** [comment]
Yep, it sucks. Do you think this should be fixed in Rails instead?

**Comment by mmitchellg5** [comment]
We're running into this as well, with Rails 4.2.4

**Comment by variousred** [comment]
FWIW as a test I deleted
```
      # Rails 4.2 goes into an infinite loop if opts[:script_name] is unset
      if (Rails::VERSION::MAJOR >= 4) && (Rails::VERSION::MINOR >= 2)
        opts[:script_name] = (config.relative_url_root if config.respond_to?(:relative_url_root))
      else
        if config.respond_to?(:relative_url_root) && config.relative_url_root.present?
          opts[:script_name] = config.relative_url_root
        end
      end
``` and I did NOT get an infinite loop, everything worked.

**Comment by mmitchellg5** [comment]
@jvdp @josevalim

**Comment by jvdp** [comment]
@mmitchellg5 hey if that fixes it, great, but doesn't that break a configuration option?

For completeness I think #3738 is a duplicate of this (and #3739 a proposed fix.)

**Comment by variousred** [comment]
@jvdp sure does, but you can just delete the conditional since apparently rails 4.2.4 does NOT go into an infinite loop without the setting.

**Comment by stanhu** [comment]
I think #3738 may be similar but not exactly the same. I just tried disabling the setting of `script_name`, and it didn't solve my use case. [Warden passes the entire `fullpath`](https://github.com/hassox/warden/blob/master/lib/warden/manager.rb#L131), leaving it to the callback function to sort things out.

**Comment by kont-noor** [comment]
@variousred Also no infinite loop issue in Rails 4.2.4

**Comment by cipater** [comment]
@josevalim I've submitted PR #3807 that attempts to fix both #3643 and bug it caused above. 

Despite suggestions to the contrary, there is definitely a infinite loop in cases where mounted engines change the SCRIPT_NAME env variable to be something other than intended path for a given scope. The '/admin/sidekiq' example in #3643 is a perfect example. I don't know enough about the internals, but it would seem something in 4.2 changed to take the SCRIPT_NAME env variable into account in the absence of a :script_name option being passed in.

**Comment by josevalim** [comment]
Awesome, thanks ofr the PR!

**Comment by lucasmazza** [comment]
Merged #4081 with a fix for the redirection issue - closing this for now.
